### PR TITLE
WIP: Hyperplane update

### PR DESCRIPTION
**Pull request by frheault**
# Quick description

Attempt at having our improved RADTRACT Hyperplane subdivision !

image
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10246281/
...

## Type of change

Check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Provide data, screenshots, command line to test (if relevant)

...

# Checklist

- [ ] My code follows the style guidelines of this project (run [autopep8](https://pypi.org/project/autopep8/))
- [ ] I added relevant citations to scripts, modules and functions docstrings and descriptions
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I moved all functions from the script file (except the argparser and main) to scilpy modules
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: scilpy/tractanalysis/bundle_operations.py [hunk 1]
```diff
@@ -117,8 +118,7 @@ def uniformize_bundle_sft(sft, axis=None, ref_bundle=None, swap=False):
             else:
                 # Bitwise XOR
                 if (bool(labels[tuple(sft.streamlines[i][0].astype(int))] >
-                         labels[tuple(sft.streamlines[i][-1].astype(int))])
-                        ^ bool(swap)):
+                         labels[tuple(sft.streamlines[i][-1].astype(int))])):
```
Comment (CHrlS98) [inline scilpy/tractanalysis/bundle_operations.py:121]: I trust you with the bitwise operations, but just want to make sure you didn't remove the `^ bool(swap)` part by accident?

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 2]
```diff
@@ -19,20 +34,394 @@ def min_dist_to_centroid(bundle_pts, centroid_pts, nb_pts):
     Array:
```
Comment (CHrlS98) [inline scilpy/tractanalysis/distance_to_centroid.py:34]: docstring is a bit lacking here.

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 3]
```diff
@@ -19,20 +34,394 @@ def min_dist_to_centroid(bundle_pts, centroid_pts, nb_pts):
...
-    dists, labels = np.expand_dims(
-        dists, axis=1), np.expand_dims(labels, axis=1)
+    _, labels = tree.query(bundle_pts, k=1)
+    labels = np.mod(labels, nb_pts) + 1
+
+    return labels.astype(np.uint16)
+
+
+def associate_labels(target_sft, min_label=1, max_label=20):
+    # DOCSTRING
```
Comment (CHrlS98) [inline scilpy/tractanalysis/distance_to_centroid.py:44]: docstring

**Review by CHrlS98 — COMMENTED** [review]
Complicated code I don't fully understand, but apart from a few missing or incomplete docstring I didn't come across anything shocking. Also looks like there are a lot of remaining PEP8 issues.

File: scilpy/tractanalysis/bundle_operations.py (hunk 1 above)
Comment (frheault) [reply scilpy/tractanalysis/bundle_operations.py:121]: It's ok since the if at line 98 is swapping labels, this is checking orientation again based on that swap (so we don't want to cancel the previous swap)

File: scripts/scil_bundle_label_map.py [hunk 4]
```diff
@@ -86,224 +142,199 @@ def _build_arg_parser():
...
+    # We trim the streamlines due to looping labels, so we have a new binary mask
+    binary_mask = np.zeros(labels_map.shape, dtype=np.uint8)
+    binary_mask[labels_map > 0] = 1
+
+    # We need to count blobs again, as the labels could be not contiguous
+    labelized, count = ndi.label(binary_mask)
+    unique, count = np.unique(labelized, return_counts=True)
+    ratio = count[1] / np.sum(count[1:])
+
+    if len(unique) > 2 and ratio < 0.9:
```
[the remark this answers is not in the document]
Comment (frheault) [reply scripts/scil_bundle_label_map.py:285]: 0.9 is arbitrary, but we want a vast majority of the voxels to be contiguous, otherwise it is a weird bundle

**Comment by frheault** [comment]
@AntoineTheb Last time you tried that branch, everything was fine with the input/output?

**Review by arnaudbore — DISMISSED** [review]
few comments but LGTM:
```
flake8 scilpy/tractograms/streamline_and_mask_operations.py 
scilpy/tractograms/streamline_and_mask_operations.py:634:5: F841 local variable 'orig_segment_len' is assigned to but never used
```

**Comment by frheault** [comment]
All good (I think), probably ready for merge?

**Comment by AntoineTheb** [comment]
@frheault there are still a lot of pep8 issues to fix. Also the `main` can probably be split into multiple functions to make the code much easier to read and review.

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 5]
```diff
@@ -1,38 +1,536 @@
...
 from scipy.spatial import KDTree
+from scipy.spatial.distance import pdist, squareform
+from sklearn.svm import SVC
+
+from scilpy.tractanalysis.streamlines_metrics import compute_tract_counts_map
+from scilpy.tractograms.streamline_operations import \
+    resample_streamlines_num_points, resample_streamlines_step_size
 
 
 def min_dist_to_centroid(bundle_pts, centroid_pts, nb_pts):
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:22]: Does this function have the right name ? It seems to return labels, should it be called something that is related to labels ?

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 6]
```diff
@@ -1,38 +1,536 @@
...
+                                  nb_pts=nb_pts)
+    logging.debug('Computed labels using the euclidian method '
+                  f'in {round(time.time() - timer, 3)} seconds')
+    min_label, max_label = labels.min(), labels.max()
+
+    if method == 'centerline':
+        labels_map = np.zeros(binary_mask.shape, dtype=np.uint16)
+        labels_map[tuple(indices.T)] = labels
+    elif method == 'hyperplane':
+        min_label, max_label = labels.min(), labels.max()
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:458]: Repeated from line 452 ?

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 7]
```diff
@@ -1,38 +1,536 @@
...
+        svc = SVC(C=1.0, kernel='rbf', random_state=1)
+
+        svc.fit(X=scaled_streamline_data, y=labels)
+        logging.debug('\tSVC fit of training data in '
+                      f'{round(time.time() - mini_timer, 3)} seconds')
+
+        # Scale the coordinates of the voxels
+        mini_timer = time.time()
+        masked_binary_mask = np.zeros(binary_mask.shape, dtype=np.uint8)
+        masked_binary_mask[::2, ::2, ::2] = binary_mask[::2, ::2, ::2]
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:498]: Why the every-two indexing ?

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 8]
```diff
@@ -1,38 +1,536 @@
...
+        voxel_coords = np.array(np.where(masked_binary_mask)).T
+        scaled_voxel_coords = scaler.transform(voxel_coords)
+
+        # Predict the labels for the voxels
+        labels = svc.predict(X=scaled_voxel_coords)
+        logging.debug('\tSVC prediction of labels in '
+                      f'{round(time.time() - mini_timer, 3)} seconds')
+
+        logging.debug('Computed labels using the hyperplane method '
+                      f'in {round(time.time() - timer, 3)} seconds')
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:508]: There is still some processing being done after. Perharps move this timer to line 519 ?

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 9]
```diff
@@ -1,38 +1,536 @@
...
+        labels_map[np.where(masked_binary_mask)] = labels
+
+        missing_indices = np.argwhere(binary_mask - masked_binary_mask)
+        valid_indices = np.argwhere(masked_binary_mask)
 
-    labels = np.mod(labels, nb_pts)
+        kd_tree = KDTree(valid_indices)
+        nn_indices = kd_tree.query(missing_indices, k=1)[1]
+        labels_map[tuple(missing_indices.T)] = \
+            labels_map[tuple(valid_indices[nn_indices].T)]
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:518]: Need comments above to explain what is going on.

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 10]
```diff
@@ -1,38 +1,536 @@
...
 
-    sum_dist = np.expand_dims(np.sum(dists, axis=1), axis=1)
-    weights = np.exp(-dists / sum_dist)
+    # Correct the labels jump to prevent discontinuities
+    if fix_jumps:
+        print('Correcting labels jump...')
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:522]: Remove print or replace with logging

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 11]
```diff
@@ -1,38 +1,536 @@
...
+        timer = time.time()
+        tmp_sft = resample_streamlines_step_size(sft, 1.0)
+        labels_map = correct_labels_jump(labels_map, tmp_sft.streamlines,
+                                         nb_pts - 2)
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:526]: Why -2 ?

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 12]
```diff
@@ -1,38 +1,536 @@
...
+        A 3D array representing the labels map.
+    """
+    sft.to_vox()
+    sft_centroid.to_vox()
+    sft.to_corner()
+    sft_centroid.to_corner()
+
+    # This allows to have a more uniform (in size) first and last labels
+    endpoints_extended = False
+    if method == 'hyperplane' and nb_pts >= 5:
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:438]: Needs more explanation, feels a bit arbitrary

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 13]
```diff
@@ -1,38 +1,536 @@
...
+        logging.debug('Corrected labels jump in '
+                      f'{round(time.time() - timer, 3)} seconds')
 
-    votes = []
-    for i in range(len(bundle_pts)):
-        vote = np.bincount(labels[i], weights=weights[i])
-        total = np.arange(np.amax(labels[i])+1)
-        winner = total[np.argmax(vote)]
-        votes.append(winner)
+    if endpoints_extended:
```
Comment (AntoineTheb) [inline scilpy/tractanalysis/distance_to_centroid.py:530]: Needs more comments to explain the process

File: scripts/scil_bundle_label_map.py [hunk 14]
```diff
@@ -86,224 +149,205 @@ def _build_arg_parser():
 def main():
```
Comment (AntoineTheb) [inline scripts/scil_bundle_label_map.py:149]: It would be best to split this function into multiple and send them into the proper modules.

File: scripts/scil_bundle_label_map.py [hunk 15]
```diff
@@ -86,224 +149,205 @@ def _build_arg_parser():
...
+            sft_list[i].data_per_streamline = {}
+            sft_list[i].data_per_point = {}
+
         if len(sft_list[i]):
             concat_sft += sft_list[i]
 
+    # Use later to trim the streamlines without assignement
+    min_len = np.min(length(concat_sft.streamlines))
+    logging.debug(
+        f'Chop bundle(s) in {round(time.time() - timer, 3)} seconds.')
```
Comment (AntoineTheb) [inline scripts/scil_bundle_label_map.py:267]: Chop ?

File: scripts/scil_bundle_label_map.py [hunk 16]
```diff
@@ -86,224 +149,205 @@ def _build_arg_parser():
...
+        # Iterate through each type to save the files
+        for basename, map in data_dict.items():
+            nib.save(nib.Nifti1Image((binary_list[i] * map), sft_list[0].affine),
+                     os.path.join(sub_out_dir, f'{basename}_map.nii.gz'))
+
+            if basename == 'correlation' and len(args.in_bundles) == 1:
+                continue
+
+            if len(sft):
+                tmp_data = ndi.map_coordinates(
```
Comment (AntoineTheb) [inline scripts/scil_bundle_label_map.py:334]: Same comment, careful with map_coordinates. Everything under the if needs more comments and could be moved into a function as well.

**Review by AntoineTheb — CHANGES_REQUESTED** [review]
Round of review. Mostly needs more comments to explain what the code is doing and why, and needs more cleanup. Functions need to be split into smaller chunks and pep8 issues need to be fixed.

File: scripts/scil_bundle_label_map.py [hunk 17]
```diff
@@ -2,47 +2,93 @@
...
 
 Formerly: scil_compute_bundle_voxel_label_map.py
+
+Author:
+-------
+Francois Rheault
+francois.m.rheault@usherbrooke.ca
+
+------------------------------------------------------------------------------------------
+Reference:
```
Comment (CHrlS98) [inline scripts/scil_bundle_label_map.py:56]: We usually put the references as EPILOG in argparse.

**Review by CHrlS98 — DISMISSED** [review]
Thanks for the extensive documentation. Single comment on the references in your main script below.

File: scripts/scil_bundle_label_map.py (hunk 17 above)
Comment (frheault) [reply scripts/scil_bundle_label_map.py:56]: @arnaudbore What is the way now?
Because now it seems that EPILOG is used for version_string
(from scilpy.version import version_string)

**Review by arnaudbore — DISMISSED** [review]

**Review by arnaudbore — APPROVED** [review]

File: scilpy/tractanalysis/distance_to_centroid.py [hunk 18]
```diff
@@ -1,38 +1,536 @@
...
+        The streamlines used to compute the labels map.
+    nb_pts (int):
+        Number of points to use for computing barycenters.
+
+    Returns:
+    -------
+    ndarray: A 3D array representing the corrected labels map.
+    """
+
+    labels_data = ndi.map_coordinates(labels_map, streamlines._data.T - 0.5,
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/tractanalysis/distance_to_centroid.py:325]: @arnaudbore @frheault we should keep this in mind as it will come back to bite us

**Review by AntoineTheb — APPROVED** [review]
LGTM
